### PR TITLE
Fix the race in podautoscaler test

### LINKED ISSUE #23361: Data race in pkg/controller/podautoscaler
[Test artifacts](https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/23302/kubernetes-pull-test-unit-integration/19096/).

I could reproduce this after 50 runs on my local machine. It may only show up with go 1.6.
```
=== RUN   TestEventCreated
I0322 14:10:33.495829   32653 horizontal.go:120] Starting HPA Controller
I0322 14:10:33.497904   32653 horizontal.go:337] Successfull rescale of test-hpa, old size: 1, new size: 2, reason: CPU utilization above target
W0322 14:10:33.498785   32653 horizontal.go:101] Failed to reconcile test-hpa: failed to compute desired number of replicas based on CPU utilization for replicationcontrollers/test-namespace/test-rc: failed to get CPU utilization: failed to get CPU consumption and request: some pods do not have request for cpu
==================
WARNING: DATA RACE
Read by goroutine 48:
  runtime.convT2E()
      /usr/local/go/src/runtime/iface.go:128 +0x0
  k8s.io/kubernetes/pkg/controller/podautoscaler.(*testCase).verifyResults()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:325 +0x2ef
  k8s.io/kubernetes/pkg/controller/podautoscaler.(*testCase).runTest()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:342 +0x723
  k8s.io/kubernetes/pkg/controller/podautoscaler.TestEventCreated()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:667 +0x20e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Previous write by goroutine 52:
  k8s.io/kubernetes/pkg/controller/podautoscaler.(*testCase).prepareTestClient.func11()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:311 +0x4c8
  k8s.io/kubernetes/pkg/client/testing/core.(*SimpleReactor).React()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/testing/core/fixture.go:271 +0x7c
  k8s.io/kubernetes/pkg/client/testing/core.(*Fake).Invokes()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/testing/core/fake.go:124 +0x36f
  k8s.io/kubernetes/pkg/client/typed/generated/core/unversioned/fake.(*FakeEvents).CreateWithEventNamespace()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/typed/generated/core/unversioned/fake/fake_event_expansion.go:31 +0x3c9
  k8s.io/kubernetes/pkg/client/typed/generated/core/unversioned.(*EventSinkImpl).Create()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/typed/generated/core/unversioned/event_expansion.go:148 +0x6e
  k8s.io/kubernetes/pkg/client/record.recordEvent()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:174 +0x215
  k8s.io/kubernetes/pkg/client/record.recordToSink()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:131 +0x1ef
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartRecordingToSink.func1()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:113 +0x5f
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher.func1()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:232 +0x14e

Goroutine 48 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:515 +0x11d
  main.main()
      k8s.io/kubernetes/pkg/controller/podautoscaler/_test/_testmain.go:98 +0x210

Goroutine 52 (running) created at:
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartEventWatcher()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:234 +0x8f
  k8s.io/kubernetes/pkg/client/record.(*eventBroadcasterImpl).StartRecordingToSink()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/client/record/event.go:114 +0x9b3
  k8s.io/kubernetes/pkg/controller/podautoscaler.NewHorizontalController()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:70 +0x140
  k8s.io/kubernetes/pkg/controller/podautoscaler.(*testCase).runTest()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:332 +0x60c
  k8s.io/kubernetes/pkg/controller/podautoscaler.TestEventCreated()
      /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:667 +0x20e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc
==================
--- PASS: TestEventCreated (12.00s)
```

On a related note, `time.Sleep` in a unit test is a huge red flag.

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #23361, using locks to prevent race.

### CODE REVIEW

**Comment by krousey** [comment]
This is insufficient. Closing the channel only signals the controller to stop. You can't be sure it's stopped. This requires more invasive changes to the entire controller framework to be able to pass a signal back that it is stopped.

**Comment by caesarxuchao** [comment]
Right. Closing the channel is not sufficient.

I changed to another approach: use the locks to protect the variables that may subject to a race. PTAL. Thanks.

File: pkg/controller/podautoscaler/horizontal_test.go [hunk 1]
```diff
@@ -321,6 +333,8 @@ func (tc *testCase) prepareTestClient(t *testing.T) *fake.Clientset {
 }
 
 func (tc *testCase) verifyResults(t *testing.T) {
+	defer tc.tcLock.RUnlock()
+	tc.tcLock.RLock()
```
Comment (krousey) [inline pkg/controller/podautoscaler/horizontal_test.go]: please invert the order of these 2 statements.

Comment (caesarxuchao) [inline pkg/controller/podautoscaler/horizontal_test.go]: Done.

**Comment by krousey** [comment]
If it passes, LGTM.

**Comment by caesarxuchao** [comment]
Adding the label per comment.

**Comment by krousey** [comment]
Closing in favor of https://github.com/kubernetes/kubernetes/pull/24537

**Pull request closed without merging**